### PR TITLE
lv_disp: add `clean_dcache_cb` callback

### PR DESCRIPTION
**Pull request by embeddedt**

### CODE REVIEW

File: src/lv_gpu/lv_gpu_stm32_dma2d.c [hunk 1]
```diff
@@ -186,7 +186,7 @@ void lv_gpu_stm32_dma2d_copy(lv_color_t * buf, lv_coord_t buf_w, const lv_color_
 void lv_gpu_stm32_dma2d_blend(lv_color_t * buf, lv_coord_t buf_w, const lv_color_t * map, lv_opa_t opa,
                               lv_coord_t map_w, lv_coord_t copy_w, lv_coord_t copy_h)
 {
-    invalidate_cache();
+    lv_disp_invalidate_cache(NULL);
```
Comment (kisvegabor) [inline src/lv_gpu/lv_gpu_stm32_dma2d.c]: It should invalidate on the being refreshed display, not the default. I suggest
```c
lv_disp_invalidate_cache(_lv_refr_get_disp_refreshing());
```

File: src/lv_gpu/lv_gpu_stm32_dma2d.c [hunk 2]
```diff
@@ -214,15 +214,6 @@ void lv_gpu_stm32_dma2d_blend(lv_color_t * buf, lv_coord_t buf_w, const lv_color
  *   STATIC FUNCTIONS
  **********************/
 
-static void invalidate_cache(void)
```
Comment (kisvegabor) [inline src/lv_gpu/lv_gpu_stm32_dma2d.c:217]: If we delete this function the user needs to take care to register a cache cleaner in the display driver.

What about keeping this function and call the custom cleaner if present else  `SCB_CleanInvalidateDCache`?

File: src/lv_core/lv_disp.c [hunk 3]
```diff
@@ -352,6 +352,21 @@ void lv_disp_trig_activity(lv_disp_t * disp)
     disp->last_activity_time = lv_tick_get();
 }
 
+/**
+ * Invalidate any CPU cache that is related to the display.
+ * @param disp pointer to an display (NULL to use the default display)
+ */
+void lv_disp_invalidate_cache(lv_disp_t * disp)
```
Comment (kisvegabor) [inline src/lv_core/lv_disp.c]: For consistence can we name `lv_disp_invalidate_cache` and `clean_dcache_cb` more similarly?

**Review by kisvegabor — COMMENTED** [review]
Hi, 
thank you!

I found a few think to discuss (see the comments)

And one more: I think `lv_disp_invalidate_cache` should be called before calling the `flush_cb`. What do you think?

File: src/lv_gpu/lv_gpu_stm32_dma2d.c (hunk 2 above)
Comment (embeddedt) [reply src/lv_gpu/lv_gpu_stm32_dma2d.c:217]: That would work, except that there'll be linking errors if `SCB_CleanInvalidateDCache` isn't available in the application, which means people would have to implement a fake stub to satisfy the linker.

File: src/lv_core/lv_disp.c (hunk 3 above)
Comment (embeddedt) [reply src/lv_core/lv_disp.c]: I've renamed it to `lv_disp_clean_dcache`.

File: src/lv_gpu/lv_gpu_stm32_dma2d.c (hunk 1 above)
Comment (embeddedt) [reply src/lv_gpu/lv_gpu_stm32_dma2d.c]: Fixed in my latest commit.

File: src/lv_gpu/lv_gpu_stm32_dma2d.c (hunk 2 above)
Comment (kisvegabor) [reply src/lv_gpu/lv_gpu_stm32_dma2d.c:217]: `SCB_CleanInvalidateDCache` is a CMSIS function, so I think we can assume that it's there.

If not - for whatever reasons - users still have a workaround.

**Comment by kisvegabor** [comment]
Thanks!

What about this question:
> And one more: I think lv_disp_invalidate_cache should be called before calling the flush_cb.

**Comment by embeddedt** [comment]
> lv_disp_invalidate_cache should be called before calling the flush_cb.

I feel like that behavior is too specific. If `flush_cb` uses the CPU to copy from the VDB to cached memory, the cache needs to be cleaned *after* that transaction, not before.

On the other hand, if a DMA transaction is used, then the cache should be cleaned before the transaction, as you suggested.

This essentially boils down to whether we should:

 * assume that one type of transaction will be used and deal with the cache in `lv_refr_vdb_flush` appropriately, or
 * leave it to the user to manage the cache in `flush_cb` themselves.

I'd rather pick the second option as it gives them the most flexibility, especially if there's another scenario down the road we didn't envision (where flushing the cache beforehand could cause other issues).

The other solution is to clean the cache both before and after `flush_cb`. That will, however, result in unnecessary data being dumped from the cache.

File: src/lv_gpu/lv_gpu_stm32_dma2d.c (hunk 2 above)
Comment (embeddedt) [reply src/lv_gpu/lv_gpu_stm32_dma2d.c:217]: Okay; I've updated it to work the way you suggested.

**Comment by kisvegabor** [comment]
I see and agree. Thanks.

Only two minor things: could you update the CHANGELOG and the docs? :slightly_smiling_face:

**Comment by embeddedt** [comment]
Documentation updated in https://github.com/lvgl/docs/commit/fa1cf0fee3472942e1b94108b65f701c6b262a26.

CHANGELOG updated in https://github.com/lvgl/lvgl/pull/1743/commits/350269c3d43e05f1ea3d1a184684a5ef0b909c6c.

Let me know if anything else needs fixing.

**Comment by kisvegabor** [comment]
Perfect, thank you!

**Pull request merged**